### PR TITLE
(maint) Add xbps used by voidlinux as a package provider

### PR DESCRIPTION
**Pull request by Animeshz**
I wanted to use puppet on void-linux, so this PR adds the package provider for xbps.

Let me know if any changes has to be made :)

### CODE REVIEW

**Comment by Animeshz** [comment]
Forgot to add strings to locales, updated with squash.

**Comment by Animeshz** [comment]
Reworded the commit message to pass `bundle exec rake commits`.

**Comment by Animeshz** [comment]
@joshcooper @tvpartytonight (just pinging).

Rebased on tip of main branch removing conflicts, can someone please review.

File: lib/puppet/provider/package/xbps.rb [hunk 1]
```diff
@@ -0,0 +1,128 @@
+require_relative "../../../puppet/provider/package"
+require "set"
+require "uri"
```
Comment (joshcooper) [inline lib/puppet/provider/package/xbps.rb:3]: I think set and uri are unused and can be removed

File: lib/puppet/provider/package/xbps.rb [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+  This provider supports the `install_options` attribute, which allows command-line flags to be passed to xbps-install.
+  These options should be specified as an array where each element is either a string or a hash."
+
+  commands :xbps_install => "/usr/bin/xbps-install"
+  commands :xbps_remove => "/usr/bin/xbps-remove"
+  commands :xbps_query => "/usr/bin/xbps-query"
+  commands :xbps_pkgdb => "/usr/bin/xbps-pkgdb"
+
+  confine :operatingsystem => [:void]
```
Comment (joshcooper) [inline lib/puppet/provider/package/xbps.rb:16]:
```suggestion
  confine 'os.name' => :void
```

File: lib/puppet/provider/package/xbps.rb [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  defaultfor :operatingsystem => [:void]
```
Comment (joshcooper) [inline lib/puppet/provider/package/xbps.rb:17]:
```suggestion
  defaultfor 'os.name' => :void
```

File: lib/puppet/provider/package/xbps.rb [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  has_feature :install_options, :uninstall_options, :upgradeable, :holdable, :virtual_packages
```
Comment (joshcooper) [inline lib/puppet/provider/package/xbps.rb:18]: I'm not familiar with `void` but is it possible to install a specific package version? If so, then you'll want `:versionable` too: https://github.com/puppetlabs/puppet/blob/7ca202bd2faf876e1904a8a2e891c10c2cee6ffc/lib/puppet/type/package.rb#L51-L54

File: locales/puppet.pot [hunk 5]
```diff
@@ -6725,6 +6725,14 @@ msgstr ""
...
 msgstr ""
 
+#: ../lib/puppet/provider/package/xbps.rb:36
+msgid "Failed to match line '%{line}"
+msgstr ""
+
+#: ../lib/puppet/provider/package/xbps.rb:43
+msgid "Error getting installed packages"
+msgstr ""
+
```
Comment (joshcooper) [inline locales/puppet.pot:6735]: You can undo this part as this file is automatically generated from source.

**Comment by joshcooper** [comment]
Hi @Animeshz thanks for your pull request! I had a few minor comments, but generally looked great.

File: lib/puppet/provider/package/xbps.rb (hunk 2 above)
Comment (Animeshz) [reply lib/puppet/provider/package/xbps.rb:16]: Is this a recent change? I wasn't aware of this syntax 👀

File: lib/puppet/provider/package/xbps.rb (hunk 4 above)
Comment (Animeshz) [reply lib/puppet/provider/package/xbps.rb:18]: Well, not actually. Its a rolling release just like arch family and repository keeps cleaning itself as new version releases. They do make sure stability among packages and update revisions on same version if necessary but it doesn't let you select the older version.

Older version remains in void-packages repository tho, which can be built on the machine from source but that's a completely different matter as binary package repository isn't involved.

File: locales/puppet.pot (hunk 5 above)
Comment (Animeshz) [reply locales/puppet.pot:6735]: I'm not sure what this means, is it that it'll be generated automatically by CI/Actions in the future?

**Comment by Animeshz** [comment]
> I had a few minor comments, but generally looked great.

I got those addressed (and tested basic functionality on local machine), let me know if anything else has to be done.

File: lib/puppet/provider/package/xbps.rb (hunk 2 above)
Comment (joshcooper) [reply lib/puppet/provider/package/xbps.rb]: yeah, in puppet 8 we no longer rely on legacy facts https://github.com/puppetlabs/puppet/wiki/Puppet-8-Compatibility#legacy-facts

File: locales/puppet.pot (hunk 5 above)
Comment (joshcooper) [reply locales/puppet.pot]: Correct, the `puppet.pot` file shouldn't be modified in this PR. The file is autogenerated in our CI pipelines.

**Comment by Animeshz** [comment]
Sorry I didn't performed lint checks. I've addressed those nuances now.

And tests should now pass. Please have a look :smile:

**Comment by joshcooper** [comment]
closing and reopening to trigger checks

**Comment by joshcooper** [comment]
@Animeshz can you fix the rubocop checks by running `bundle exec rubocop -a` and squash your commits?

**Comment by Animeshz** [comment]
It used to pass all checks that time, seems like rubocop rules have been changed since then.

24-03-13-12h10m36s

Let me rebase, fix and squash up real quick.

File: locales/puppet.pot [hunk 6]
```diff
@@ -6741,6 +6741,14 @@ msgstr ""
...
```
Comment (joshcooper) [inline locales/puppet.pot:6751]: Please remove these changes as the file is autogenerated.

File: lib/puppet/provider/package/xbps.rb [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    # Search for virtual package
+    output = xbps_query("-Rs", resource_name).chomp
+
+    # xbps-query -Rs output is '[*] package-name-version     description'
+    regex = /^\[\*\]+\s(\S+)-(\S+)\s+\S+/
+    match = regex.match(output)
+
+    return nil unless match
+
+    { name: match.captures[0], ensures: match.captures[1], provider: self.class.name }
```
Comment (joshcooper) [inline lib/puppet/provider/package/xbps.rb:105]:
```suggestion
    { name: match.captures[0], ensure: match.captures[1], provider: self.class.name }
```

File: lib/puppet/provider/package/xbps.rb [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    cmd = %w[-R -y]
+    cmd += uninstall_options if @resource[:uninstall_options]
+    cmd << resource_name
+
+    xbps_remove(*cmd)
+  end
+
+  # The latest version of a given package
+  def latest
+    query&.[] :ensures
```
Comment (joshcooper) [inline lib/puppet/provider/package/xbps.rb:82]:
```suggestion
    query&.[] :ensure
```

File: lib/puppet/provider/package/xbps.rb [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+  # Fetch the list of packages that are currently installed on the system.
+  def self.instances
+    packages = []
+    execpipe([command(:xbps_query), "-l"]) do |pipe|
+      # xbps-query -l output is 'ii package-name-version     desc'
+      regex = /^\S+\s(\S+)-(\S+)\s+\S+/
+      pipe.each_line do |line|
+        match = regex.match(line.chomp)
+        if match
+          packages << new({ name: match.captures[0], ensures: match.captures[1], provider: name })
```
Comment (joshcooper) [inline lib/puppet/provider/package/xbps.rb:33]:
```suggestion
          packages << new({ name: match.captures[0], ensure: match.captures[1], provider: name })
```

**Review by joshcooper — APPROVED** [review]
Thanks @Animeshz this will be released in the next puppet 8 version

**Pull request merged**